### PR TITLE
Added a new custom code inspection to insist on the use of OkDialog rather than RunUI to dismiss dialogs in tests

### PR DESCRIPTION
**Pull request by bspratt**
Unlike RunUI, OkDialog waits for a dialog to actually go away which avoids potential race conditions in tests.

It's true that in most cases there were accompanying calls to WaitForClosedForm safeguarding against race conditions, but there were also enough places without such safeguards that adopting OkDialog as our consistent and enforceable coding style is worth a bit of churn here.

Also updated the developer Wiki with notes about this.

Internal test issue, no reporting user

### CODE REVIEW

**Review by brendanx67 — APPROVED** [review]
Looks like a lot of improvement to me. I definitely was not aware of how pervasive the more verbose form of RunUI();WaitForClosedForm(); had become. Definitely better to standardize on OkDialog() as guaranteeing the combination.

I did notice a couple cases where:
1. There was still a WaitForClosedForm() immediately after OkDialog()
2. There was a dlg.OkDialog() inside a RunUI(() => {block}); with multiple lines and then a WaitForClosedForm()

File: pwiz_tools/Skyline/TestPerf/SmallMolLibrariesTutorialTest.cs [hunk 1]
```diff
@@ -187,7 +187,7 @@ protected override void DoTest()
             var confirmAddDlg = ShowDialog<AlertDlg>(viewLibUI.AddAllPeptides);
             using (new CheckDocumentState(1, 34, 38, 246))
             {
-                RunUI(confirmAddDlg.OkDialog);
+                OkDialog(confirmAddDlg, confirmAddDlg.OkDialog);
```
[the remark this answers is not in the document]
Comment (bspratt) [reply pwiz_tools/Skyline/TestPerf/SmallMolLibrariesTutorialTest.cs:190]: Yes, there were quite a number of places where other things would ensure the complete dismissal of a dialog before the test proceeded, but as you say it's just as well to settle on a standard.

File: pwiz_tools/Skyline/TestUtil/TestFunctional.cs [hunk 2]
```diff
@@ -2232,8 +2232,7 @@ private static void AddMod(string uniModName, bool isVariable, EditListDlg<Setti
             });
             WaitForClosedForm(addStaticModDlg);
```
[the remark this answers is not in the document]
Comment (bspratt) [reply pwiz_tools/Skyline/TestUtil/TestFunctional.cs:2233]: No, it won't pick up multi-line cases like this. That would require a pretty complicated parser. I suppose we could add another inspection for any uses of blah.OkDialog or blah.CancelDialog which aren't embedded in a call to OkDialog().

[the remark this answers is not in the document]
Comment (brendanx67) [reply pwiz_tools/Skyline/TestUtil/TestFunctional.cs:2233]: These could also be converted to RunDlg() which also includes the wait. The benefit of using a separate call to OkDialog() is that it provides a place to put a PauseTest() to see the form state before the button to submit it is clicked. I wonder if we could consider entirely getting rid of WaitForClosedForm() in our tests. At least without a comment on the line to disable static analysis complaining.

[the remark this answers is not in the document]
Comment (bspratt) [reply pwiz_tools/Skyline/TestUtil/TestFunctional.cs:2233]: Yes, if we're going all the way with this - or with Nick's alternative - we should kill off WaitForClosedForm(), largely if not completely.

[the remark this answers is not in the document]
Comment (brendanx67) [reply pwiz_tools/Skyline/TestUtil/TestFunctional.cs:2233]: Nick's PR is not an alternative to this. In fact, it currently has no impact at all on the most common construction in our tests:

var dlg = ShowDialog<T>(FunctionThatShowsAForm);
RunUI(() => dlg.DoSomething());
OkDialog(dlg, dlg.OkDialog);

It works with:

 RunDlg<T>(FunctionThatShowsAForm, dlg => {
  dlg.DoSomething();
  dlg.OkDialog();
});

And adds a new function RunLogDialog<T>() which I don't totally understand yet, but does not provide a direct replacement for the first construct above.

**Comment by bspratt** [comment]
This has been sitting awhile - there were a couple of suggestions for polishing up some code, but as an inspection this is good to go AFAIK. Any reason not to merge? Or do we really want to wait till those suggestions are implemented?

**Comment by brendanx67** [comment]
I think it is fine. Along with Nick's changes to RunDlg, we definitely made
some progress in this area.

**Pull request merged**